model/test_results_iterator.go

```diff
@@ -0,0 +1,98 @@
+package model
+
+import (
+	"context"
+	"io/ioutil"
+
+	"github.com/evergreen-ci/pail"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"go.mongodb.org/mongo-driver/bson"
+)
+
+type TestResultsIterator interface {
+	Iterator
+	// Item returns the current TestResult item held by the iterator.
+	Item() TestResult
+}
+
+type testResultsIterator struct {
+	bucket      pail.Bucket
+	iter        pail.BucketIterator
+	currentItem TestResult
+	exhausted   bool
+	closed      bool
+	catcher     grip.Catcher
+}
+
+// NewTestResultsIterator returns an iterator
+func NewTestResultsIterator(bucket pail.Bucket) TestResultsIterator {
+	return &testResultsIterator{
+		bucket:  bucket,
+		catcher: grip.NewBasicCatcher(),
+	}
+}
+
+func (i *testResultsIterator) Next(ctx context.Context) bool {
+	if i.exhausted || i.closed {
+		return false
+	}
+
+	if i.iter == nil {
+		iter, err := i.bucket.List(ctx, "")
+		if err != nil {
+			i.catcher.Wrap(err, "listing bucket contents")
+			return false
+		}
+		i.iter = iter
+	}
+	if !i.iter.Next(ctx) {
```

Comment:
We should probably add and errors from `i.iter` to `i.catcher`

Reply:
I added the errors after `Next()` is called.